src/modules/adaptors/Morpho/MorphoAaveV2ATokenAdaptor.sol

```diff
@@ -0,0 +1,175 @@
+            address aToken = abi.decode(adaptorData, (address));
+            return _balanceOfInUnderlying(aToken, msg.sender);
+        }
+    }
+
+    /**
+     * @notice Returns the cellars balance of the positions aToken.
+     */
+    function balanceOf(bytes memory adaptorData) public view override returns (uint256) {
+        address aToken = abi.decode(adaptorData, (address));
+        return _balanceOfInUnderlying(aToken, msg.sender);
+    }
+
+    /**
+     * @notice Returns the positions aToken underlying asset.
+     */
+    function assetOf(bytes memory adaptorData) public view override returns (ERC20) {
+        IAaveToken token = IAaveToken(abi.decode(adaptorData, (address)));
+        return ERC20(token.UNDERLYING_ASSET_ADDRESS());
+    }
+
+    /**
+     * @notice This adaptor returns collateral, and not debt.
+     */
+    function isDebt() public pure override returns (bool) {
+        return false;
+    }
+
+    //============================================ Strategist Functions ===========================================
+    /**
+     * @notice Allows strategists to lend assets on Morpho.
+     * @dev Uses `_maxAvailable` helper function, see BaseAdaptor.sol
+     * @param aToken the aToken to lend on Morpho
+     * @param amountToDeposit the amount of `tokenToDeposit` to lend on Morpho.
+     */
+    function depositToAaveV2Morpho(IAaveToken aToken, uint256 amountToDeposit) public {
+        ERC20 underlying = ERC20(aToken.UNDERLYING_ASSET_ADDRESS());
+        amountToDeposit = _maxAvailable(underlying, amountToDeposit);
+        underlying.safeApprove(address(morpho()), amountToDeposit);
+        morpho().supply(address(aToken), amountToDeposit);
+
+        // Zero out approvals if necessary.
+        _revokeExternalApproval(underlying, address(morpho()));
+    }
+
+    /**
+     * @notice Allows strategists to withdraw assets from Morpho.
+     * @param aToken the atoken to withdraw from Morpho.
+     * @param amountToWithdraw the amount of `tokenToWithdraw` to withdraw from Morpho
+     */
+    function withdrawFromAaveV2Morpho(IAaveToken aToken, uint256 amountToWithdraw) public {
+        morpho().withdraw(address(aToken), amountToWithdraw, address(this));
+    }
+
+    /**
+     * @notice Returns the balance in underlying of collateral.
+     * @param poolToken the Aave V2 a Token user has supplied as collateral
+     * @param user the address of the user to query their balance of
+     */
+    function _balanceOfInUnderlying(address poolToken, address user) internal view returns (uint256) {
+        (uint256 inP2P, uint256 onPool) = morpho().supplyBalanceInOf(poolToken, user);
+
+        uint256 balanceInUnderlying;
+        // Morpho indexes are scaled by 27 decimals, so divide by 1e27.
+        if (inP2P > 0) balanceInUnderlying = inP2P.mulDivDown(morpho().p2pSupplyIndex(poolToken), 1e27);
+        if (onPool > 0) balanceInUnderlying += onPool.mulDivDown(morpho().poolIndexes(poolToken).poolSupplyIndex, 1e27);
```

Comment:
Bear in mind that these indexes may be outdated (they are updated at each interaction with the protocol and thus are not guaranteed to be up-to-date). If you want to access the true balance in underlying, you need to use the indexes calculated by the Lens (or calculate them yourself)

Reply:
Ah I see I'll take a look at the lens contract, it is okay if these numbers are slightly off, like in our Compound V2 code we use the stored exchange rate mantissa, and that is fine because it never substantially deviates from the actual value. So if that is the same case here, I am not worried about it.